Tweaks and improvements

stormtek opened the pull request:
Added support for an array of values as well as a list of values to be sent to range()
Added support for a width parameter to be set on the input field
Improved handling of layout of slider options
Moved styling for slider to it's own stylesheet to make it easier to use in other projects
Improved readme documentation on how to use the slider

Commit:
Simplified and extended example html
Split css so that styling for the slider is in a separate file
Tweaked js to allow an array of options to be passed in if desired

stormtek committed:
Update README.md

stormtek committed:
Update README.md

Commit:
Split css for the slider itself into a separate file
Added support for a width value to be added to the input field
Improved layout of options

stormtek committed:
Update README.md

jcemer commented:
Your examples are broken http://stormtek.github.io/zepto-range/

stormtek commented:
Broken in what way?

jcemer commented:
![image](https://f.cloud.github.com/assets/353504/672953/8fa3934a-d8a0-11e2-8067-141782108151.png)

stormtek commented:
it looks fine for me ...

stormtek commented:
![zepto](https://f.cloud.github.com/assets/1317080/672958/01e21684-d8a1-11e2-87ca-c12e95954b99.png)

jcemer commented:
Try to open on chrome, firefox... I'm using Chrome Version 27.0.1453.110 on Mac.

stormtek commented:
im using ff on windows and it looks fine

stormtek commented:
It looks fine in Firefox on Ubuntu 12.10, Firefox on OSX, but you are right ... it breaks on chrome on both OSX and windows ...

stormtek commented:
So a styling issue with chrome? Does your original demo work on Chrome?

stormtek commented:
That is weird, I will look into it further. Cos that is a serious issue ...

stormtek commented:
Thanks for the heads up by the way :)

Ruffio commented:
also crome on android is brøken. it used to work.
Den 19/06/2013 07.33 skrev "stormtek" notifications@github.com:

> So a styling issue with chrome? Does your original demo work on Chrome?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jcemer/zepto-range/pull/2#issuecomment-19663950
> .

stormtek commented:
So it it a problem with chrome on almost all platforms ... it does work on firefox on all those platforms though

stormtek commented:
Looking at the generated html, the wrapper div being created has width being set to 0px when we are grabbing the width automatically rather than taking it from the input field value. So we are not getting the general width from the browser for some reason. That makes it a bug in the javascript. Hopefully that should not be too hard to fix.

ricardobeat commented:
@stormtek you are trying to get `this.input.attr('width')` where it should be `this.input.width()`. That's one of the problems.

stormtek commented:
That is needed to allow the user to define the width of the slider in the input field. The problem is that if that is not specified then it is returning 0 in Chrome. I think I can fix it fairly easily by dropping to this.input.width() if we get 0 from the existing call.

stormtek committed:
Fixed bug in slider width that showed up in Chrome

stormtek commented:
How do I get that change back to the link we are looking at?

stormtek commented:
That does fix the problem when I run it locally in Chrome on my macbook though

stormtek commented:
Ok, may just need to wait a few minutes for testing live. But I do believe that is fixed now.

stormtek commented:
Yea, that is up and working in Chrome on both osx and windows.

Ruffio commented:
the test seams to work now in chrome om android. it is not supposed to fill
100% right?
Den 19/06/2013 08.01 skrev "stormtek" notifications@github.com:

> Yea, that is up and working in Chrome on both osx and windows.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jcemer/zepto-range/pull/2#issuecomment-19664614
> .

stormtek committed:
Update index.html

stormtek commented:
If there is no wrapper div with a width specified and no width set on the input then it should fill 100%. I have updated the demo to show that with one of the sliders.

Ruffio commented:
beautiful :) you should check description for max parameter. and what does
inside do? missing a description?
Den 19/06/2013 11.26 skrev "stormtek" notifications@github.com:

> If there is no wrapper div with a width specified and no width set on the
> input then it should fill 100%. I have updated the demo to show that with
> one of the sliders.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jcemer/zepto-range/pull/2#issuecomment-19672053
> .

stormtek commented:
what do you mean by check description for max parameter?

Ruffio commented:
The description for max parameter at the index page is incorrect. The value
should be lager than min valmue, not larger by itself.
Den 20/06/2013 03.13 skrev "stormtek" notifications@github.com:

> what do you mean by check description for max parameter?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jcemer/zepto-range/pull/2#issuecomment-19725470
> .

stormtek commented:
I see it now ... stupid eyes filling in what I expect to be there and glossing over what is there ...

stormtek committed:
Update README.md

Ruffio commented:
How do you set initial value?
Is it possible to get 'end value' while the moving is occuring?

Perhaps describe this in readme.md?

stormtek committed:
Update README.md

stormtek commented:
I do not believe it is possible to get the 'end value' while the slider is in motion, though I could be wrong. To be honest, I cannot think of a sensible time where you would want to anyway.

stormtek committed:
Now displays min and max values if no options are passed to range()

stormtek committed:
Update README.md

stormtek commented:
I have now adjusted the slider to show the min and max values that are specified in the input field if no options are passed to range(). This makes it more obvious what is going on, and also makes the slider usable even if semantic information is not being presented through custom labels. It also makes it more obvious that the value is being set correctly when the slider is created.

stormtek committed:
Update README.md

Ruffio commented:
Would it be possible to implement an option to disable legend/bars?

If the range is large and the witdh is small one will get 'one' (what appears to be) one lang white bar below the slider.

It would be nice, if one could only disable the bars, but still keeping min and max legend.

stormtek commented:
Is that so that you could use it to simulate a continuous slider? In theory that should not be too hard, though it will require mucking with the js a bit. It sounds like an interesting idea though, so I will have a play round with it and see.

Ruffio commented:
Exactly.

Here is a testpage: http://mbrevis.com/default.aspx

I want to refresh the background color of a div, when the value of range is changed. On slow mobiles many calls to 'UpdateColor' javascript function can be a heavy workload, therefore I was interested in 'end-value' in order to offload workload.

Still not figured out how to set the range value after is has been initialized :-(

Hope it makes sense...

stormtek commented:
Is that so that you could use it to simulate a continuous slider? In theory that should not be too hard, though it will require mucking with the js a bit. It sounds like an interesting idea though, so I will have a play round with it and see.

Commit:
Updated to allow parameter to be passed to input to allow hiding of lines for empty values, updated demo to show this on a slider

Commit:
Merge branch 'gh-pages' of https://github.com/stormtek/zepto-range into gh-pages

stormtek committed:
Update README.md

stormtek commented:
I have added a parameter that allows you to hide empty labels if you wish. It is as easy as adding

showEmptyLabels="false"

to your input field.

stormtek commented:
I also just noticed that with a large number of entries the styling starts to fall apart with a large number of input. In fact, I think that the more inputs the worse it will get. That is annoying ...

stormtek commented:
Setting the initial value is part of the initialization. You need to set

value="x"

as one of the parameters for the input field. I believe that no value set for this will set the slider to the minimum value.

stormtek commented:
You do need to trigger a reload of the page though, not just a refresh

Commit:
Improved styling, still doesnt handle extreme cases perfectly though

stormtek committed:
Update README.md

Ruffio commented:
Later today I'll try to implement all your changes, thanks for that :-)

Reload of page is not an option in my case, but I think that I'll try to remove the input and the add it again. Would that work?

stormtek commented:
I'm just looking at the js, and I think there must be a way to get at the method which sets the value of the slider - which would be a useful thing to be able to get anyway. I will see what I can do ... :)

Commit:
Added method to allow us to change the value of the slider dynamically through code, added a button to one of the sliders in the example to show it working.

stormtek commented:
There we go, a method added to allow for dynamic change of value. I will update the readme now

Ruffio commented:
Damn, you are fast. Can't wait til tonight when I have the chance to implement all the changes.

Thanks alot :-)

jcemer commented:
Please, dont use that as a chat. This is not the purpose.

stormtek committed:
Update README.md

stormtek commented:
Sorry. Where would be the best place to start said chat though?

Ruffio commented:
Sorry. I have requested some features/pointed out some 'missing' (in my view obvious) elements. Would you rather have that I made bug incidents? No mater what, I still appreciate the work that have been done, and I think this is the right place to say that.

jcemer commented on index.html:
```diff
@@ -3,53 +3,86 @@
+    <link rel="stylesheet" href="styles.css"/>
+    <link rel="stylesheet" href="zepto-range.css"/>
+    <link href="//fonts.googleapis.com/css?family=Ubuntu:300,700" rel="stylesheet" type="text/css"/>
+    <script src="zepto.js"></script>
```
You cant open a pull request changing everything. By the way, put scripts on _head_ is not a good practice.

jcemer commented:
This pull request changes everything. I never will accept this. If you prefer, keep going with discussion.

jcemer commented:
Reconsidering. I think you can split this in many PR and I will appreciate to help with all. Thanks

stormtek committed:
Update README.md

stormtek commented:
I never intended to go out and 'change everything'. I had found a couple of places where this slider did not quite work when we started to put it in the real world and thought that you might appreciate the input. We could easily have decided not to contribute those adjustments. In the process of issuing this pull request a number of other things have come up.

And I do not feel that I have 'changed everything' anyway. The core functionality of the slider is still there untouched. And I think that what you have done with it is incredible. Overall the slider is very nice to work with and there has been very little hassle. At no point have I tried to break that or change it in any significant way. I have just been trying to add in some more useful functionality and extend the documentation into something useful.

Commit:
Restyled page to be centred again

Commit:
Fixed styling for mobile (which I had broken)

Ruffio commented:
Yesterday I did some testing and overall it seems to be working very nicely :-)

I do have some smaller hick ups but that must be some of my local setup that is playing with me. F.x. if I set the value to 2 then the slider moves to a position below 0 (range = 0-255) and the coloured part of the slider is not moving :-( Your sample page works perfectly so it must be my setup that is causing the problem. In 12 hours I'll have to test again and hopefully find my bug. I'll report back with my findings.(http://mbrevis.com/default.aspx)

Ruffio commented:
I have done some testing and can confirm, that the above mentioned hickup is also occurring at your sample page. In FF or Chrome click on the handle and try to move it to a position before/left of 'start'/minimum. You will notice that you can get the handle a little bit out of range before it jumps back again. This bug doesn't occur at jcemer's sample page.

In Firebug I can see, that the little jump is 20px, does that give a clue?

Another very little thing. When setting the value by pressing the button (Set value to 2), then first the coloured bar goes to the correct value and the handle goes to the value. This is really a minor thing and I believe that I'm the only one noticing this.

Commit:
If value is not specified then we now set default value to min

stormtek committed:
Update README.md

Commit:
Fixed jitter bug where it was possible to get the handle on the slider off either end of the slider briefly

stormtek commented:
That jittery bug is now fixed

stormtek commented:
You can notice the coloured bar moving before the handle if you click on a value part way down a slider too (this is triggering the same code as the button click does (through setValue())). It only happens the first time that slider is interacted with though. I don't know that there is anything we can do about it - I have a feeling it is the css property transform() (which is set on the handle) that takes some extra time to process ...

Commit:
Fixed glitch where changing value would have handle lag a long way behind fill

stormtek commented:
It turns out that the lag of the handle (e.g. when clicking button first time) is because the css transition is taking longer. However, it is possible to wait for a transitionend event (slighlty different for different vendors) and adjust the width of the fill once that event fires. A little bit of fiddling was then needed to make sure that we only wait in certain situations (so not when the user is moving the slider in 'normal' usage). The end result is the removal of that lag (apart from a little bit of lag once the event has fired, but that doesn't look any where near as bad, i think being caused by the transition: easing set on .range .fill in the css).

Tested on Firefox (OSX and Windows, Ubuntu), iOS simulator, Chrome (OSX, Windows), Safari on OSX

stormtek commented:
And I just noticed that Safari on OSX is not rendering most of the divs on my test page correctly ... I think it is to do with the width not being set correctly ... This needs looking into :(

Commit:
Fixed bug with width being set to NaN if not specified which broke slider in some browsers

stormtek commented:
Fixed the safari bug

Ruffio commented:
Have you updated your sample page with the new code? Jumping is still occuring there...

stormtek commented:
Every time I commit it is updated. Make sure you reload page properly - sometimes the javascript can get cached. It is not jumping for me.

Ruffio commented:
My fault. I can detect no jumping :-) Great.

The lag of the coloured part of the bar (left of handle) is 'interesting'. You are waiting for handle animation to complete, before starting the fill animation, correct? Howcome this has not been an issue with jcemer's version?

Ruffio commented:
Can't you 'just' set the fill width to the handle position when the handle is moving? Instead of animating the fill too?

Ruffio commented:
Can I fork your PR?

This is working better for me. Now the fill bar follows the handle exactly:

 //    if (waitForTransitionEnd) {
        //        var me = this;
        //        var eventTriggered = false;
        //        this.btn[0].addEventListener('transitionend', function (event) {
        //            if (!eventTriggered) {
        //                me.finishMove(to, pos);
        //                eventTriggered = true;
        //            }
        //        }, false);
        //        this.btn[0].addEventListener('webkitTransitionEnd', function (event) {
        //            if (!eventTriggered) {
        //                me.finishMove(to, pos);
        //                eventTriggered = true;
        //            }
        //        }, false);
        //        this.btn[0].addEventListener('oTransitionEnd', function (event) {
        //            if (!eventTriggered) {
        //                me.finishMove(to, pos);
        //                eventTriggered = true;
        //            }
        //        }, false);
        //    } else {
        //        this.finishMove(to, pos);
        //    }
            $.translateX(this.btn[0], pos);
        //},
        //finishMove: function (to, pos) {
            this.fill.width(pos);
            this.input.trigger('move', [to, this]);
        },

Ruffio commented:
Would it be possible to reinitialize the range or use percent instead of px? 

This comes in handy if you use the range component on a mobile device, that users some times tend to rotate, thereby changing the width...

I have tried to use  $('#Name').range(); but that doesn't reset the width as I had hoped :-(

stormtek committed:
Tweak to make the fill work properly with the handle

stormtek committed:
Updated styling for handling fill better with handle

stormtek commented:
I have updated the fill to only animate when we want it to. This mostly fixes how the fill bar follows the handle. The original issue (click part way down a slider and the fill gets there way before the handle lots of the time) is an issue with both my version and the original one put up by jcemer. This seems to be an issue with the css transition taking a while sometimes (which we cannot control). So we need to wait until that transition has ended before shifting the fill bar. The fill bar is literally just having the width changed, no animation, in this case, but that obviously takes just long enough to produce the weird flicker. It is almost like the browser is sometimes smoothing the width adjustment slightly. Anyway, overall it performs better like this - the glitch is a lot less obvious. If you can find a way that removes all flicker in the adjustment of fill and handle (in all scenarios) then let me know. This is the best performance I could squeeze out of it. And the majority of the time users should be dragging the slider anyway (not clicking part way), so they will not notice the issue.

stormtek commented:
$('#id').range() won't reset the width since the range element already exists. I guess it would be possible to have a reload method (potentially with some parameters?) that would remove the extra divs etc that were added, along with the js reference to the range (it is storing a list of of initialized range sliders), and then recreate it using range(). In theory this should work, but it would require a new js method to handle all of that.

stormtek commented:
I am not sure if we can set the width of an element in % rather than pixels. It is worth a look. Then again, I'm not sure if it will fix the problem of a browser not liking sub pixel widths. It is worth a shot though.

stormtek commented:
Yea, using percentages doesn't help. We are still at the mercy of the browser as to how a sub pixel value will round. We need to be able to guarantee that the sum of all the widths is just less than the width of the slider. Thus we need an algorithm which specifies how many values to round up and how many to round down and then evenly space those values along the slider.

Examples:

slider width: 100 px
64 values
desired width of each value: ~1.5px
round half up to 2px and half down to 1px
32 \* 1 + 32 \* 2 = 32 + 64 = 96px - just under the 100px for our slider

slider width: 100px
80 values
desired width of each value: 1.25px
round 3/4 down to 1px and 1/4 up to 2px
60 \* 1 + 20 \* 2 = 60 + 40 = 100px

So I believe that we can use the decimal value to give the desired ratio of up:down rounding and use that to space the widths evenly along the slider. All we need to do then is get each label and set its width appropriately.

Commit:
Wrote algorithm to handle adjusting label widths for sub pixel values. This fixes a layout bug that showed up in safari where all the widths were set to 0. It takes slightly longer to generate the slider now, but we are no longer at the mercy of the browser to interpret how to round values. This is particularly important for sliders with lots of values in a smaller space.

Commit:
Added minified js for zepto library

stormtek committed:
Update README.md

stormtek committed:
Update index.html

stormtek committed:
Update README.md

stormtek commented:
For some reason the minified js is suppressing some touch events on mobile ... not sure why

Commit:
Updated zepto.min.js to include the touch module

stormtek committed:
Update README.md

stormtek committed:
Switched back to zepto.min.js now that I have included the touch module

Ruffio commented:
My experience with minified js is that it does matter which algorithm/site
that is used:(
Den 27/06/2013 05.47 skrev "stormtek" notifications@github.com:

> For some reason the minified js is suppressing some touch events on mobile
> ... not sure why
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jcemer/zepto-range/pull/2#issuecomment-20095720
> .

stormtek commented:
It turns out it was a problem with the minified js. I dug a little deeper and found that the version of zepto.js I had included the touch module but the version of zepto.min.js I had did not. I pulled down the zepto source and rebuilt both with the touch module in. So it works again now :)

stormtek commented:
I can now confirm that on Android touch is working well for my version in:

Chrome, Firefox, Opera Mobile, Dolphin, and the stock Internet browser

It is also behaving in Safari on the iOS simulator on my macbook. The slider also behaves in our native iOS app that uses phonegap.

Commit:
Fixed small layout issue in creation of labels where widths were slightly wrong

Commit:
Merge branch 'gh-pages' of https://github.com/stormtek/zepto-range into gh-pages

Commit:
Removed wait for transition stuff, since overall it was not looking quite as nice, now it seems to be behaving ...

Commit:
Added the ability to easily specfiy a custom fill colour for the slider - whether at init time or at runtime

Commit:
Adjusted layout of slider slightly to include a 10px buffer at either end. This allows us to then detect a click on the legend near either end of the slider and move it to the min or max value. This provides better usability for the slider.

Commit:
Fixed bug in setValue() where we were setting the value to the index, rather than the correct underlying value

stormtek committed:
Update README.md

Commit:
remove unneccessary css

Commit:
Merge branch 'gh-pages' of https://github.com/stormtek/zepto-range into gh-pages

Commit:
Fix handling of click just off ends of slider to work with touch

Commit:
Styling tweak to improve performance of clicking off end of slider when using on mobile

Commit:
Small error check addition

Commit:
Added a function to set the width of a slider to a specified value once it has been created. Updated demo to show this in action.

stormtek committed:
Update README.md

Commit:
Tweaked algorithm for setting label widths, it should be a bit more robust now

Commit:
Merge branch 'gh-pages' of https://github.com/stormtek/zepto-range into gh-pages

stormtek committed:
Update README.md

Commit:
Added function to get value from range. Fixed bug where clicking off bottom of slider set value wrong.

Commit:
Merge branch 'gh-pages' of https://github.com/stormtek/zepto-range into gh-pages

stormtek committed:
Update README.md

Commit:
Added output of current value for one of the sliders based on event handler

stormtek committed:
Update README.md

Commit:
Tweaked positioning of some buttons

Commit:
Fixed bug in set width

Commit:
Fixed algorithm for setting widths of labels

Commit:
Fixed small bug in removeRange() where i was being used rather than index

stormtek committed:
Set input to hidden rather than offscreen

We were having some weird scrolling when the element got focus that was fixed by changing this

stormtek committed:
Added support for two new parameters

showFill: allows a user to specify whether or not to show the fill
rtl: allows a user to specify whether or not to use an rtl layout on the fill

stormtek committed:
Update README.md
